mitiq/cdr/clifford_training_data.py

```diff
@@ -0,0 +1,514 @@
+
+    elif method_replace == 'probabilistic':
+        rz_non_cliff_selected[2, :] = _probabilistic_angle_to_clifford(
+            rz_non_cliff_selected[2, :], sigma_replace)
+
+    else:
+        raise Exception(
+            'method_replace must = "closest", "random", "probabilistic"')
+    # to store replaced angles:
+    angles_replaced = rz_non_cliff_selected[2, :].copy()
+    # Add back into rest of data and re-order instructions:
+    new_circ = np.column_stack((all_cliff, rz_non_cliff_selected))
+    new_circ = np.column_stack((new_circ, rz_non_cliff_copy))
+    i = np.argsort(new_circ[0, :])
+    projected_circuit_data = new_circ[:, i]
+    projected_circuit = _array_to_circuit(
+        projected_circuit_data, empty_circuit_copy)
+
+    return projected_circuit, angles_original, angles_replaced
+
+
+def count_non_cliffords(
+    circuit: Circuit,
+) -> float:
+    """Function to check how many non-Clifford gates are in a give circuit.
+
+    Args:
+        circuit: some quantum circuit already decomposed into the basis
+                 {Rz, Rx(pi/2), CNOT, X}
+
+    Returns:
+        number of non-Clifford gates in the given circuit.
+    """
+    data, _ = _circuit_to_array(circuit)
+    mask_rz = data[1, :] == 'rz'
+    rz_circ_data = data[:, mask_rz]
+    mask_non_cliff = _is_clifford_angle(rz_circ_data[2, :])
+    mask_non_cliff = ~mask_non_cliff
+    rz_non_cliff = rz_circ_data[:, mask_non_cliff]
+    return len(rz_non_cliff[0])
+
+
+def _circuit_to_array(
+    circuit: Circuit
+) -> (np.ndarray, Circuit):
+    """Function to return the order of gates, their names and paramters in a
+       more managable data structure than a Qiskit
+    quantum circuit.
+
+    Args:
+        circ (Circuit): cirq circuit (decomposed).
+
+    Returns:
+        data (np.ndarray): np.array([order], [names], [parameters], [qubits])
+                           where order is the order of the gates from 0 to
+                           depth, names are the names of the gates,
+                           parameters are the paramters specifying the
+                           gates and qubits and cbits are the qubits and
+                           classical bits on which they act.
+        Circuit: empty circuit with same qubit layout as original.
+    """
+    order = []
+    gates_list = []
+    qubits_list = []
+    operation_list = []
+    parameters_list = []
+    operations = circuit.all_operations()
+    # loop through and construct arrays:
+    for i, operation in enumerate(operations):
+        operation_list.append(operation)
+        order.append(i)
+        qubits = operation.qubits
+        gate = operation.gate
+        if isinstance(gate, cirq.ops.common_gates.CXPowGate):
+            qubit = [qubits[0], qubits[1]]
+            parameters = None
+            name = 'cx'
+            gates_list.append(name)
+            qubits_list.append(qubit)
+            # cbit_list.append(None)
+            parameters_list.append(parameters)
```

Comment:
I'm still confused to why this "circuit as array" data structure is necessary. Here all of the information stored - `qubit`, `parameters`, etc. - is contained in the cirq.Circuit data structure, except the newly introduced string name. Can we drop the name and just deal directly with the Cirq circuit?